packages/@aws-cdk/pipelines/lib/codepipeline/codepipeline-engine.ts

```diff
@@ -0,0 +1,888 @@
+import { DockerCredential, dockerCredentialsInstallCommands, DockerCredentialUsage } from '../docker-credentials';
+import { GraphNode, GraphNodeCollection, isGraph, AGraphNode, PipelineGraph } from '../helpers-internal';
+import { BuildDeploymentOptions, IDeploymentEngine } from '../main/engine';
+import { appOf, assemblyBuilderOf, embeddedAsmPath, obtainScope } from '../private/construct-internals';
+import { toPosixPath } from '../private/fs';
+import { enumerate, flatten, maybeSuffix } from '../private/javascript';
+import { writeTemplateConfiguration } from '../private/template-configuration';
+import { CodeBuildFactory, mergeCodeBuildOptions, stackVariableNamespace } from './_codebuild-factory';
+import { ArtifactMap } from './artifact-map';
+import { CodeBuildStep } from './codebuild-step';
+import { CodePipelineActionFactoryResult, ICodePipelineActionFactory } from './codepipeline-action-factory';
+
+/**
+ * Creation properties for a `CodePipelineEngine`
+ */
+export interface CodePipelineEngineProps {
+  /**
+   * The name of the CodePipeline pipeline
+   *
+   * @default - Automatically generated
+   */
+  readonly pipelineName?: string;
+
+  /**
+   * Create KMS keys for the artifact buckets, allowing cross-account deployments
+   *
+   * The artifact buckets have to be encrypted to support deploying CDK apps to
+   * another account, so if you want to do that or want to have your artifact
+   * buckets encrypted, be sure to set this value to `true`.
+   *
+   * Be aware there is a cost associated with maintaining the KMS keys.
+   *
+   * @default false
+   */
+  readonly crossAccountKeys?: boolean;
+
+  /**
+   * CDK CLI version to use in self-mutation and asset publishing steps
+   *
+   * If you want to lock the CDK CLI version used in the pipeline, by steps
+   * that are automatically generated for you, specify the version here.
+   *
+   * You should not typically need to specify this value.
+   *
+   * @default - Latest version
+   */
+  readonly cdkCliVersion?: string;
+
+  /**
+   * Whether the pipeline will update itself
+   *
+   * This needs to be set to `true` to allow the pipeline to reconfigure
+   * itself when assets or stages are being added to it, and `true` is the
+   * recommended setting.
+   *
+   * You can temporarily set this to `false` while you are iterating
+   * on the pipeline itself and prefer to deploy changes using `cdk deploy`.
+   *
+   * @default true
+   */
+  readonly selfMutation?: boolean;
+
+  /**
+   * Set if the pipeline itself builds Docker container assets
+   *
+   * NOTE: this only applies to Docker assets used for the Pipeline
+   * or the Pipeline stack itself. It does not apply for Docker assets
+   * used in the Stages and Stacks that are *deployed* by this pipeline.
+   *
+   * Configures privileged mode for the self-mutation CodeBuild action.
+   *
+   * @default false
```

Comment:
Can we make this a smart default?

Reply:
No, for 2 reasons:

- We don't have a cloud assembly yet to inspect. 
- This actually needs to be turned on, deployed, and only THEN can you add the dependency on Docker (otherwise the initial SelfUpdate step will fail for lack of a `privileged: true` setting).

So if we tie it together, you lose the ability to turn it on explicitly.

Reply:
OK. I guess it's not a very common use case any way, right?